# Zapytania HTTP i REST API - wprowadzenie

Jednym ze sposobów na komunikację z siecią/serwerami z poziomu skryptu napisanego w Pythonie
jest wykonywanie zapytań HTTP. Pozwala nam to zarówno na pobieranie danych z serwerów jak też na ich modyfikację
i dodawanie nowych informacji.

Wyróżniamy kilka podstawowych rodzajów zapytań HTTP:
- **GET** - pobranie danych
- **POST** - dodanie nowych danych
- **PUT**- aktualizacja istniejących danych
- **DELETE** - usunięcie danych

Współcześnie komunikacja z serwerami odbywa się za pośrednictwem tzw. **REST API**.
W dużym skrócie wygląda to w taki sposób, że wysyłamy zapytanie do serwera, a w odpowiedzi otrzymujemy
obiekt typu **JSON**.

Format **JSON** jest powszechnie wykorzystywany i w większości zastąpił format XML.
W formacie JSON obiekty przypominają obiekty w językach programowania.

Do wykonywania zapytań HTTP skorzystamy z biblioteki *requests*:

In [1]:
!pip install requests


[notice] A new release of pip is available: 24.0 -> 24.3.1
[notice] To update, run: python.exe -m pip install --upgrade pip


Najpierw importujemy bibliotekę

In [2]:
import requests

Teraz możemy wykonać proste zapytanie do serwera i zapisać odpowiedź w zmiennej *response*.

Dla przykładu skorzystamy ze strony http://api.open-notify.org która udostępnia API pozwalające pobrać
informacje o osobach przebywających obecnie w kosmosie, a także o położeniu ISS.

In [3]:
response = requests.get("http://api.open-notify.org/astros.json")

Odpowiedź z serwera składa się z kilku składowych. Jedną z nich jest *status code*.

In [4]:
print(response.status_code)

200


Kod ten informuje nas o statusie zwróconym przez nasze zapytanie. Kod **200** oznacza, że zapytanie
przebiegło pomyślnie.

Dla przykładu spróbujmy wykonać zapytanie pod nieistniejący adres.

In [5]:
response = requests.get("http://api.open-notify.org/non-existing")
print(response.status_code)

404


Otrzymaliśmy kod **404**. Oznacza to, że nie znaleziono na serwerze żądanego zasobu.

Wyróżniamy kilka grup kodów odpowiedzi:
- **1xx** - informacyjne
- **2xx** - sukces
- **3xx** - przekierowanie
- **4xx** - błąd klienta
- **5xx** - błąd serwera

Więcej informacji: https://pl.wikipedia.org/wiki/Kod_odpowiedzi_HTTP

Powtórzmy zapytanie i wypiszmy jego odpowiedź JSON.

In [6]:
response = requests.get("http://api.open-notify.org/astros.json")
print(response.json())

{'people': [{'craft': 'ISS', 'name': 'Oleg Kononenko'}, {'craft': 'ISS', 'name': 'Nikolai Chub'}, {'craft': 'ISS', 'name': 'Tracy Caldwell Dyson'}, {'craft': 'ISS', 'name': 'Matthew Dominick'}, {'craft': 'ISS', 'name': 'Michael Barratt'}, {'craft': 'ISS', 'name': 'Jeanette Epps'}, {'craft': 'ISS', 'name': 'Alexander Grebenkin'}, {'craft': 'ISS', 'name': 'Butch Wilmore'}, {'craft': 'ISS', 'name': 'Sunita Williams'}, {'craft': 'Tiangong', 'name': 'Li Guangsu'}, {'craft': 'Tiangong', 'name': 'Li Cong'}, {'craft': 'Tiangong', 'name': 'Ye Guangfu'}], 'number': 12, 'message': 'success'}


Dostaliśmy bogatą odpowiedź od serwera - listę osób przebywających w kosmosie.
Chcielibyśmy je wyświetlić w czytelniejszej formie. Do tego celu wykorzystamy wbudowaną bibliotekę **json**.

In [7]:
import json

def jprint(obj):
    text = json.dumps(obj, sort_keys=True, indent=4)
    print(text)

Teraz możemy użyć naszej funkcji do czytelnego wyświetlenia otrzymanej odpowiedzi.

In [8]:
jprint(response.json())

{
    "message": "success",
    "number": 12,
    "people": [
        {
            "craft": "ISS",
            "name": "Oleg Kononenko"
        },
        {
            "craft": "ISS",
            "name": "Nikolai Chub"
        },
        {
            "craft": "ISS",
            "name": "Tracy Caldwell Dyson"
        },
        {
            "craft": "ISS",
            "name": "Matthew Dominick"
        },
        {
            "craft": "ISS",
            "name": "Michael Barratt"
        },
        {
            "craft": "ISS",
            "name": "Jeanette Epps"
        },
        {
            "craft": "ISS",
            "name": "Alexander Grebenkin"
        },
        {
            "craft": "ISS",
            "name": "Butch Wilmore"
        },
        {
            "craft": "ISS",
            "name": "Sunita Williams"
        },
        {
            "craft": "Tiangong",
            "name": "Li Guangsu"
        },
        {
            "craft": "Tiangong",
            "name": "Li

Jak można zauważyć, odpowiedź jest w formie słownika. Możemy łatwo przeiterować po wszystkich osobach
przebywających w kosmosie.

In [9]:
people = response.json()["people"]
for person in people:
    print(person["name"])

Oleg Kononenko
Nikolai Chub
Tracy Caldwell Dyson
Matthew Dominick
Michael Barratt
Jeanette Epps
Alexander Grebenkin
Butch Wilmore
Sunita Williams
Li Guangsu
Li Cong
Ye Guangfu
